<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<link rel="stylesheet" type="text/css" href="config_generatorStyle.css">
<title>BioLockJ</title>
</head>

<body>
<!--file button-->
<div class="nav-bar">
  <div class="dropdown">
    <button class="dropbtn" onclick="projectOptions()">Projects</button>
    <div class="dropdown-content" id="projects">
      <a onclick="newProj()" href="#">New</a>
      <a onclick="openProjDisplay()" href="#">Open local config file</a>
      <a id="recent" href="#">Recent</a>
    </div>
  </div>
</div>

<div class="banner">BioLockJ</div>

<div id="openConfig" class="menu">
  <form>
    Select BiolockJ configuration file
    <input type="file" name="localFile" id="localFile">
    <button type="button" id="submitLocalFile" onclick="loadLocalFile()">Submit</button>
  </form>
</div>

<!--options for control panel-->
<div class="menu" id='mainMenu'>
  <div class="tab">
    <button id="defaultOpen" class="tablinks" onclick="openTab(event, 'projectTab')">Project</button>
    <button class="tablinks" onclick="openTab(event, 'moduleTab')">Modules</button>
    <button class="tablinks" onclick="openTab(event, 'inputTab')">Input</button>
    <button class="tablinks" onclick="openTab(event, 'clusterTab')">Cluster</button>
    <button class="tablinks" onclick="openTab(event, 'scriptTab')">Script</button>
    <button class="tablinks" onclick="openTab(event, 'reportTab')">Report</button>
    <button class="tablinks" onclick="openTab(event, 'mailTab')">E-mail</button>
    <button class="tablinks" onclick="openTab(event, 'rTab')">R</button>
    <button class="tablinks" onclick="openTab(event, 'exeTab')">Exe</button>
    <button class="tablinks" onclick="openTab(event, 'rdpTab')">RDP</button>
    <button class="tablinks" onclick="openTab(event, 'qiimeTab')">QIIME</button>
    <button class="tablinks" onclick="openTab(event, 'krakenTab')">Kraken</button>
    <button class="tablinks" onclick="openTab(event, 'slimmTab')">SLIMM</button>
  </div>
  <!--tabbed menu-->

  <div id="projectTab" class="tabcontent">
    <!--<h3>Project</h3>-->
    <form id="project">
      <p>Configuration file name<input type="text" name="configFile" required></p>
      <p>Project name:
        <input type="text" name="name" required><br></p>
      <p>Root directory: <input type="text" name="rootDir" required><br>
        <h5>(The root directory is the parent directory for BioLockJ project run directories)</h5></p>
      <p>Copy input files into the parent directory:<br>
        <input type="radio" name="copyInput" value="Y" required> Yes<br>
        <input type="radio" name="copyInput" value="N"> No<br></p>
      <p>Delete temporary files:<br>
        <input type="radio" name="deleteTempFiles" value="Y" required> Yes<br>
        <input type="radio" name="deleteTempFiles" value="N"> No<br></p>
      <!--  -->
      <p>Path to default properties file:
        <input type="text" name="defaultProps"><br> add file picker here
        <i>(Path to a default BioLockJ configuration file containing default
        property values that are overridden if defined in the primary
        configuration file.)</i></p>
      <p>Log level: <br><i>Determines log level sensitivity.</i><br>
        <input type="radio" name="logLevel" value="DEBUG">DEBUG<br>
        <input type="radio" name="logLevel" value="INFO">INFO<br>
        <input type="radio" name="logLevel" value="WARN">WARN<br>
        <input type="radio" name="logLevel" value="ERROR">ERROR<br></p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type='button' class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="moduleTab" class="tabcontent">
    <button type="button" onclick="modulesToCurrentConfig()">Save</button>
    <button type="button" class="nextTab">Save and move to next tab</button>
    <fieldset>
      <legend>Module options</legend>
      <div id="module">
        <span> <center> Drag module options to reorder them.<br>
          Only highlighted (green) modules will be added to your config file. <br> </center>
        </span>
        <ul id='module_ul' style="padding-left: 5px">
        </ul>
      </div>
      <div id='moduleInfo'>
        <center>Hover mouse over module for more information.</center><br>
        <div id='moduleInfoDiv'>
          <div>Info will display here!</div>
          <!--inner div for adding information to-->
        </div>
      </div>
    </fieldset>
  </div>

  <div id="inputTab" class="tabcontent">
    <form id="input">
      <p>Input directory (or directories): <input type="text" name="dirs" required><br>
        <h5>(Input directory must contain files expected by the first module executed.
        Multiple directories must be comma-separated.)</h5></p>
      <p>Files in input directory to ignore: <input type="text" name="ignoreFiles"><br>
        <h5>(Multiple files must be comma-separated.)</h5></p>
      <p>Demultiplex: <br>
        <input type="radio" name="demultiplex" value="Y" required> Yes<br>
        <input type="radio" name="demultiplex" value="N"> No<br>
        <h5>(If yes, sequence files include reads from multiple samples
          and sample IDs must be extracted from the sequence headers.
          RDP and Kraken classifiers only.  If no, there is one sample per file
          and the file name must contain the sample ID.)</h5></p>
      <p>Paired reads: <br>
        <input type="radio" name="pairedReads" value="Y" required> Yes<br>
        <input type="radio" name="pairedReads" value="N"> No<br>
        <h5>(If yes, file names must include input.forwardFileSuffix or
        input.reverseFileSuffix.)<h5></p>
      <p><File name suffix to indicate a forward read:
            <input type="text" name="forwardFileSuffix"><br>
            File name suffix to indicate a reverse read.
            <input type="text" name="reverseFileSuffix"><br>
            </p>
      <p>Trim prefix:
        <input type="text" name="trimPrefix"><br>
      <h5>(For files named by Sample ID, provide the
        prefix preceding the ID to trim when extracting Sample ID.
If input.demultiplex=Y, provide any characters in the sequence header preceding the ID.
For fastq, typically “@”.)<h5></p>
      <p>Trim suffix:
        <input type="text" name="trimSuffix"><br>
        <h5>(For files named by Sample ID, provide the suffix after the ID, often
           this is just the file extension.  Do not include read direction indicators l
           isted in input.forwardFileSuffix/reverseFileSuffix.
           If input.demultiplex=Y, provide 1st character in the sequence header after ID;
           for fastq, typically “:” char.)</h5></p>
      <p>Minimum number of sequences per sample (BioLockJ will discard samples with sequences below this.):
        <input type="number" name="rarefyMin"><br></p>
      <p>Maximum number of sequences per sample (BioLockJ will discard samples with sequences above this.):
        <input type="number" name="rarefyMax"><br></p>
      <p>Path to file containing primers to trim (File must contain only one sequence per line):
        <input type="text" name="trimSeqPath"><br></p>
      <p>Metadata file path, <i>must have unique column headers</i>:
        <input type="text" name="metadata"><br></p>
      <p>Define how null values are represented in metadata: (input.nullValue)<br>
        <input type="text" name="nullValue">
      </p>
      <p>Define column delimiter for input.metadata (input.columnDelim)<br>
        <input type="text" name="columnDelim"></p>
      <p>Barcode in header?<br>
        <input type="radio" name="barcodeInHeader" value="Y"> Yes<br>
        <input type="radio" name="barcodeInHeader" value="N"> No<br></p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="clusterTab" class="tabcontent">
    <form id="cluster">
      <button onclick="saveTab('cluster')">Next</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
      <p>Cluster host address (cluster.host):<br>
        <input type="text" name="host"><br></p>
      <p>Batch command to submit jobs on the cluster: <br>
        <input type="text" name="batchCommand"><br>
      </p>
      <p>Cluster parameters: <br>
        <input type="text" name="params"><br></p>
      <p>Validate cluster parameters: <br>
        <input type="radio" name="validateParams" value="Y"> Yes<br>
        <input type="radio" name="validateParams" value="N"> No<br></p>
      <p>Cluster modules to load before execution: <br>
        <input type="text" name="modules"><br></p>
      <p>The email notification will include an scp command to download this BioModule outputDir<br>
        <input type="text" name="downloadModule"><br></p>
      <p>A directory on the local workstation where the cluster.downloadModule will be downloaded<br>
        <input type="text" name="downloadDir"><br></p>
      <p>Cluster.scriptHeader<br>
        <input type="text" name="scriptHeader"><br></p>
    </form>
  </div>


  <div id="scriptTab" class="tabcontent">
    <form id="script">
      <p>Exit program if errors in script occur: (script.exitOnError)<br>
        <input type="radio" name="exitOnError" value="Y"> Yes<br>
        <input type="radio" name="exitOnError" value="N"> No (failures will be logged to failure directory) <br></p>
      <p>Run on cluster: <br>
        <input type="radio" name="runOnCluster" value="Y"> Yes<br>
        <input type="radio" name="runOnCluster" value="N"> No</p>
      <p>Number of sequence files to process per script:<br>
        <input type="number" name="batchSize"><br></p>
      <p>Command to grant script execute permissions:
        <input type="text" name="chmodCommand" value="chmod 774"><br></p>
      <p>Number of threads:
        <input type="text" name="numThreads"><br></p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="reportTab" class="tabcontent">
    <form id="report">
      <p>Include number of hits in output: <br>
        <input type="radio" name="numHits" value="Y"> Yes<br>
        <input type="radio" name="numHits" value="N"> No<br></p>
      <p>Add number of reads to output: <br>
        <input type="radio" name="numReads" value="Y"> Yes<br>
        <input type="radio" name="numReads" value="N"> No<br></p>
      <p>Add full taxonimic names to output:<br>
        <input type="radio" name="fullTaxonomyNames" value="Y"> Yes<br>
        <input type="radio" name="fullTaxonomyNames" value="N"> No<br></p>
      <p>Add genus prefix to species name in output:<br>
        <i>(If yes, ParserModule adds genus prefix to species name in output tables)</i><br>
        <input type="radio" name="addGenusToSpeciesName" value="Y">Yes<br>
        <input type="radio" name="addGenusToSpeciesName" value="N">No
        <br></p>
      <p>Add only <strong>first initial</strong> of genus to species name in output:<br>
        <input type="radio" name="useGenusFirstInitial" value="Y">Yes<br>
        <input type="radio" name="useGenusFirstInitial" value="N">No
        <br></p>
      <p>Metadata attributes for R script to add to statistical models:<br>
        <input type="text" name="attributes">
        <br></p>
      <p>Min threshold for OTU counts (<i>below which they will be ignored</i>):<br>
        <input type="text" name="minOtuCount"><br></p>
      <!--  <p>Delimiter to sepearte genus and species:<br>
        <input type="text" name="emptySpaceDelim"><br></p>-->
      <p>Taxonomic levels included in report:<br>
        <input type="checkbox" name="taxonomyLevels" value="domain">Domain<br>
        <input type="checkbox" name="taxonomyLevels" value="phylum">Phylum<br>
        <input type="checkbox" name="taxonomyLevels" value="class">Class<br>
        <input type="checkbox" name="taxonomyLevels" value="order">Order<br>
        <input type="checkbox" name="taxonomyLevels" value="family">Family<br>
        <input type="checkbox" name="taxonomyLevels" value="genus">Genus<br>
        <input type="checkbox" name="taxonomyLevels" value="species">Species<br>
      </p>
      <p>Report log base: <br>
        <input type="radio" name="logBase" value="10">10<br>
        <input type="radio" name="logBase" value="e">e<br></p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="mailTab" class="tabcontent">
    <form id="mail">
      <p>Max size for log file attachment (in MB):
        <input type="text" name="maxAttachmentSizeMB"><br>
      </p>
      <p>Encrypted E-mail password:<br>
        <i>(Encrypted password from email.from account.  If BioLockJ is passed a 2nd parameter
          (in addition to the config file), the 2nd parameter should be the clear-text password.
          The password will be encrypted and stored in the prop file for future use.)</i><br>
        <strong>WARNING: Base64 encryption is only a trivial roadblock for malicious users.
          This functionality is intended merely to keep clear-text passwords out of the
          configuration files and should only be used with a disposable email.from account.</strong>
        <input type="text" name="encryptedPassword"><br>
        <p>Notification emails sent from this account, provided email.encryptedPassword is valid:
          <input type="text" name="from"><br>
          <p>Comma-separated email recipients list:
            <input type="text" name="to"><br>
          </p>
          <p>Email SMTP Host:<br>
            <input type="text" name="smtp.host"><br>
          </p>
          p>Email SMTP port:<br>
          <input type="text" name="smtp.port"><br>
        </p>
        <p>Set the SMTP authorization property.<br>
          <input type="radio" name="smtp.auth" value="Y">Yes<br>
          <input type="radio" name="smtp.auth" value="N">No<br></p>
        <p>Set the SMTP start TLS property.<br>
          <input type="radio" name="smtp.starttls.enable" value="Y">Yes<br>
          <input type="radio" name="smtp.starttls.enable" value="N">No<br></p>
        <button onclick="saveTab(this.parentNode.id)">Save</button>
        <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="rTab" class="tabcontent">
    <form id="r">
      <p>If Y, will generate R Script log files:<br>
        <input type="radio" name="debug" value="Y">Yes<br>
        <input type="radio" name="debug" value="N">No <br>
      </p>
      <p>R script adds nominal statistical models for the listed metadata attributes. (Comma seperated list)<br>
        <input type="text" name="nominalData"><br>
      </p>
      <p>R script adds numeric statistical models for the listed metadata attributes. (Comma seperated list)<br>
        <input type="text" name="numericData"><br>
      </p>
      <p>Rare OTU threshold: <br>
        <i>If >1, R will filter OTUs below # provided.  If <1, R will treat # as percentage and ignore OTUs not found in that percentage of table rows in each taxa level.</i>
        <input type="number" name="rareOtuThreshold"><br>
      </p>
      <p><strong>Attributes Filter</strong><br>List of attributes for r to report on:
        <input type="text" name="filterAttributes"><br>
      </p>
      <p>List of logical operators to apply to attributes filter:
        <input type="text" name="filterOperators"><br>
      </p>
      <p>Filter values: <br><i><An orrdered list of values to compare with attributes filter)</i>
        <input type="text" name="filterValues"><br>
      </p>
      <p>Rows with NA values for attributes ignored in R script:
        <input type="text" name="filterNaAttributes">
      </p>
      <p>Number of breaks in P-value histograms output by R script:
        <input type="text" name="numHistogramBreaks">
      </p>
      <p>P-value adjustment scope: Used to set the p.adjust "n" parameter for how many simultaneous p-value calculations:<br>
        <input type="radio" name="adjPvalsScope" value="GLOBAL">GLOBAL<br>
        <input type="radio" name="adjPvalsScope" value="LOCAL">LOCAL<br>
        <input type="radio" name="adjPvalsScope" value="TAXA">TAXA<br>
        <input type="radio" name="adjPvalsScope" value="ATTRIBUTE">ATTRIBUTE<br>
      </p>
      <p>This is the base color used for labels & headings in the PDF report:<br>
        <input type="text" name="baseColor">
      </p>
      <p>This color is used to highlight significant attributes:<br>
        <input type="text" name="highlightColor">
      </p>
      <p>Sets the color of scatterplot points:<br>
        <input type="text" name="pointColor">
      </p>
      <p>Sets the p.adjust "method" parameter:<br><a href=https://stat.ethz.ch/R-manual/R-devel/library/stats/html/p.adjust.html>More info</a><br>
        <input type="radio" name="pAdjMethod" value="holm">Holm<br>
        <input type="radio" name="pAdjMethod" value="hochberg">Hochberg<br>
        <input type="radio" name="pAdjMethod" value="hommel">Hommel<br>
        <input type="radio" name="pAdjMethod" value="bonferroni">Honferroni<br>
        <input type="radio" name="pAdjMethod" value="BH">Benjamini & Hochberg (BH)<br>
        <input type="radio" name="pAdjMethod" value="BY">Benjamini & Yekutieli (BY)<br>
        <input type="radio" name="pAdjMethod" value="fdr">FDR<br>
        <input type="radio" name="pAdjMethod" value="none">none<br>
      </p>
      <p>Sets the format used in R sprintf() function:<br>
        <input type="text" name="pValFormat"><br>
      </p>
      <p>Set p-value cutoff used to assign label r.highlightColor:<br>
        <input type="number" name="pvalCutoff"><br>
      </p>
      <p>Set PDF report plot label maximum width:<br>
        <input type="number" name="plotWidth"><br>
      </p>
      <p>Set PDF report plot label position: (r.topRowAdj)<br>
        <input type="number" name="topRowAdj"><br>
      </p>
      <p>Set PDF report plot label position: (r.bottomRowAdj)<br>
        <input type="number" name="bottomRowAdj"><br>
      </p>
      <p>Set PDF report plot label position: (r.leftColAdj)<br>
        <input type="number" name="leftColAdj"><br>
      </p>
      <p>Set PDF report plot label position: (r.rightColAdj)<br>
        <input type="number" name="rightColAdj"><br>
      </p>
      <p>Set the size of histogram & boxplot headings in PDF reports<br>
        <input type="number" name="headingLabelSize"><br>
      </p>
      <p>Set the size of metadata attribute names in PDF reports<br>
        <input type="number" name="attributeLabelSize"><br>
      </p>
      <p># minutes before R Script times out:<br>
        <input type="number" name="timeout"><br>
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="exeTab" class="tabcontent">
    <p>If the program executable isn't already in your classpath ($PATH), add the executable path below:
    </p>
    <!--Make it so default commands are hints that will not be saved.-->
     <form id="exe">
      <p>Classifier executable command:
        <input type="text" name="classifier">
      </p>
      <p>Classifier executable command:
        <input type="text" name="classifierParams">
      </p>
      <p>Executable RScript command:
        <input type="text" name="rScript">
      </p>
      <p>Executable Java command:
        <input type="text" name="java">
      </p>
      <p>Executable python command:
        <input type="text" name="python">
      </p>
      <p>Executable gzip command:
        <input type="text" name="gzip">
      </p>
      <p>Executable awak command:
        <input type="text" name="awak">
      </p>
      <p>Executable bowtie2 command:
        <input type="text" name="bowtie">
      </p>
      <p>Optional bowtie2 parameters:
        <input type="text" name="bowtieParams">
      </p>
      <p>Executable PEAR command:
        <input type="text" name="pear">
      </p>
      <p>Optional PEAR parameters:
        <input type="text" name="pearParams">
      </p>
      <p>Executable samtools command:
        <input type="text" name="samtools">
      </p>
      <p>Executable vsearch command:
        <input type="text" name="vsearch">
      </p>
      <p>Optional PEAR parameters:
        <input type="text" name="vsearchParams">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="rdpTab" class="tabcontent">
    <form id="rdp">
      <p>Required RDP minimum threshold score for valid OTUs:
        <input type="text" name="minThresholdScore">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>

  <div id="qiimeTab" class="tabcontent">
    <form id="qiime">
      <p>Alpha diverisity metrics: <br> Options listed online:
        <a href="http://scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html">
          scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html</a>.<br>
        <input type="text" name="alphaDiversityMetrics">
      </p>
      <p>Remove chimeras after open or de novo OTU picking:<br>
        <input type="radio" name="removeChimeras" value="Y">Yes<br>
        <input type="radio" name="removeChimeras" value="N">No <br>
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>

  <div id="krakenTab" class="tabcontent">
    <form id="kraken">
      <p>Path to Kraken database:
        <input type="text" name="kraken.db">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a id="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>

  <div id="slimmTab" class="tabcontent">
    <form id="slimm">
      <p>Path to SLIMM database:
        <input type="text" name="slimm.db">
      </p>
      <p>Path the bowtie2 reference genome index:
        <input type="text" name="refGenomeIndex">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>
</div>
<div class="banner">
  <!--Working on logo here-->
  <?xml version="1.0" standalone="No"?>
  <svg width="150" height="75" version="1.1" xmlns="http://www.w3.org/2000/svg">
    <!--rounded edges rect-->
    <rect x="60" y="20" rx="10" ry="10" width="30" height="30" stroke="black" fill="transparent" stroke-width="4"/>

    <!--<ellipse cx="75" cy="75" rx="20" ry="5" stroke="red" fill="transparent" stroke-width="5"/>-->

    <path d="M60,23 Q70,5  80,5 T90,23" fill="none" stroke="rgb(243,144,29)" stroke-width="2"/>
    <path d="M60,23 Q65,3  75,5 T90,23" fill="none" stroke="rgb(0,112,60)" stroke-width="2"/>
    <text x="69" y="43" dx="0" font-family="Times">J</text>
    <!--<path d="M70,30 L80,30  M75,30 L75,35 " fill="none" stroke="rgb(0,112,60)" stroke-width="2"/>-->
  </svg>
</div>
<script type="text/javascript" src="config_generator.js"></script>
</body>

</html>
