<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<link rel="stylesheet" type="text/css" href="config_generatorStyle.css">
<title>BioLockJ</title>
</head>

<body>
<!--file button-->
<div class="nav-bar">
  <div class="dropdown">
    <button class="dropbtn" onclick="projectOptions()">Projects</button>
    <div class="dropdown-content" id="projects">
      <a onclick="newProj()" href="#">New</a>
      <a onclick="openProjDisplay()" href="#">Open local config file</a>
      <a id="recent" href="#">Recent</a>
    </div>
  </div>
</div>

<div class="banner">BioLockJ</div>

<div id="openConfig" class="menu">
  <form>
    Select BiolockJ configuration file
    <input type="file" name="localFile" id="localFile">
    <button type="button" id="submitLocalFile" onclick="loadLocalFile()">Submit</button>
  </form>
</div>

<!--options for control panel-->
<div class="menu" id='mainMenu'>
  <div class="tab">
    <button id="defaultOpen" class="tablinks" onclick="openTab(event, 'projectTab')">Project</button>
    <button class="tablinks" onclick="openTab(event, 'moduleTab')">Modules</button>
    <button class="tablinks" onclick="openTab(event, 'inputTab')">Input</button>
    <button class="tablinks" onclick="openTab(event, 'clusterTab')">Cluster</button>
    <button class="tablinks" onclick="openTab(event, 'scriptTab')">Script</button>
    <button class="tablinks" onclick="openTab(event, 'reportTab')">Report</button>
    <button class="tablinks" onclick="openTab(event, 'mailTab')">E-mail</button>
    <button class="tablinks" onclick="openTab(event, 'rTab')">R</button>
    <button class="tablinks" onclick="openTab(event, 'exeTab')">Exe</button>
    <button class="tablinks" onclick="openTab(event, 'rdpTab')">RDP</button>
    <button class="tablinks" onclick="openTab(event, 'qiimeTab')">QIIME</button>
    <button class="tablinks" onclick="openTab(event, 'krakenTab')">Kraken</button>
    <button class="tablinks" onclick="openTab(event, 'slimmTab')">SLIMM</button>
  </div>
  <!--tabbed menu-->

  <div id="projectTab" class="tabcontent">
    <form id="project">
      <p>Configuration file name
        <input id='configFile' type="text" name="project.configFile" required></p>
      <p>Copy input files into the parent directory:<br>
        <input type="radio" name="project.copyInput" value="Y" required> Yes<br>
        <input type="radio" name="project.copyInput" value="N"> No<br></p>
      <p>Delete temporary files:<br>
        <input type="radio" name="project.deleteTempFiles" value="Y" required> Yes<br>
        <input type="radio" name="project.deleteTempFiles" value="N"> No<br></p>
      <p>Path to default properties file:
        <input type="text" name="project.defaultProps"><br> add file picker here
        <i>(Path to a default BioLockJ configuration file containing default
        property values that are overridden if defined in the primary
        configuration file.)</i></p>
      <p>Log level: <br><i>Determines log level sensitivity.</i><br>
        <input type="radio" name="logLevel" value="DEBUG">DEBUG<br>
        <input type="radio" name="logLevel" value="INFO">INFO<br>
        <input type="radio" name="logLevel" value="WARN">WARN<br>
        <input type="radio" name="logLevel" value="ERROR">ERROR<br></p>
      <p>Project runtime environment: <br>
        <input name='project.env' type="radio" value="aws"> AWS<br>
        <input name='project.env' type="radio" value="cluster"> Cluster<br>
        <input name='project.env' type="radio" value="local"> Yes<br></p>
      <p>Run project in Docker:<br>
        <input type="radio" name="project.runDocker" value="Y"> Yes<br>
        <input type="radio" name="project.runDocker" value="N"> No<br></p>
      <p>A directory on the local workstation where the cluster.downloadModule will be downloaded<br>
        <input type="text" name="project.downloadDir"><br></p>
      <p>Set # minutes before direct Java module script will time out and fail<br>
        <input type="number" name="java.timeout" value="">
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type='button' class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="moduleTab" class="tabcontent">
    <button type="button" onclick="modulesToCurrentConfig()">Save</button>
    <button type="button" class="nextTab">Save and move to next tab</button>
    <!-- TODO: Link button to currentConfig -->
    <button type='button' id='implicitOverride'>Config Implicit Override</button>
    <h3 class='hidden' id='implicitWarning'>
      Implicit modules will be added automatically by BioLockJ as needed.</h3>
    <fieldset>
      <legend>Module options</legend>
      <div id="module">
        <span> <center> Drag module options to reorder them.<br>
          Only highlighted (green) modules will be added to your config file. <br> </center>
        </span>
        <ul id='module_ul' style="padding-left: 5px">
        </ul>
      </div>
      <div id='moduleInfo'>
        <center>Hover mouse over module for more information.</center><br>
        <div id='moduleInfoDiv'>
          <div>Info will display here!</div>
          <!--inner div for adding information to-->
        </div>
      </div>
    </fieldset>
  </div>
  <div id="inputTab" class="tabcontent">
    <form id="input">
      <p>Input directory (or directories): <input type="text" name="input.dirPaths" required><br>
        <h5>(Input directory must contain files expected by the first module executed.
        Multiple directories must be comma-separated.)</h5></p>
      <p>Files in input directory to ignore: <input type="text" name="input.ignoreFiles"><br>
        <h5>(Multiple files must be comma-separated.)</h5></p>
      <p>File name suffix to indicate a forward read:
            <input type="text" name="input.suffixFw"><br>
            File name suffix to indicate a reverse read.
            <input type="text" name="input.suffixRw"><br>
            </p>
      <p>Trim prefix:
        <input type="text" name="input.trimPrefix"><br>
      <h5>(For files named by Sample ID, provide the
        prefix preceding the ID to trim when extracting Sample ID.
If input.demultiplex=Y, provide any characters in the sequence header preceding the ID.
For fastq, typically “@”.)<h5></p>
      <p>Trim suffix:
        <input type="text" name="input.trimSuffix"><br>
        <h5>(For files named by Sample ID, provide the suffix after the ID, often
           this is just the file extension.  Do not include read direction indicators l
           isted in input.forwardFileSuffix/reverseFileSuffix.
           If input.demultiplex=Y, provide 1st character in the sequence header after ID;
           for fastq, typically “:” char.)</h5></p>
      <p>Maximum accepted length of sequence read:<br>
        <input type="number" name="input.seqMaxLen"><br></p>
      <p>Minimum accepted length of sequence read:<br>
        <input type="number" name="input.seqMinLen"><br></p>
      <p>Minimum number of sequences per sample for use in rarefication (BioLockJ will discard samples with sequences below this.):<br>
        <input type="number" name="rarefier.min"><br></p> <!--these should be hidden unless rarefy module is selected (also change names)-->
      <p>Randomly select maximum number of sequences per sample use in rarefication:<br>
        <input type="number" name="rarefier.max"><br></p>
      <p>Path to file containing primers to trim (File must contain only one sequence per line):
        <input type="text" name="trimPrimers.filePath"><br></p>
      <p>Discard reads with no primers?<br>
        <input type="radio" name="trimPrimers.requirePrimer" value="Y"> Yes<br>
        <input type="radio" name="trimPrimers.requirePrimer" value="N"> No<br></p>
      <!--add demux -->
      <p>Set the Demultiplexer strategy. If using barcodes, they must be provided in the
        metadata.filePath with in column name defined by metadata.barcodeColumn.<br>
        <input type="radio" name="demux.strategy" value="barcode_in_header">barcode_in_header<br>
        <input type="radio" name="demux.strategy" value="id_in_header">id_in_header<br></p>
        <input type="radio" name="demux.strategy" value="do_not_demux.">do_not_demux.<br></p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

<!-- TODO: Add metadata tab -->
      <p>Define how null values are represented in metadata: (input.nullValue)<br>
        <input type="text" name="nullValue"></p>
      <p>Define column delimiter for input.metadata (input.columnDelim)<br>
        <input type="text" name="columnDelim"></p>


  <div id="clusterTab" class="tabcontent">
    <form id="cluster">
      <button onclick="saveTab('cluster')">Next</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
      <p>Cluster host address (cluster.host):<br>
        <input type="text" name="cluster.host"><br></p>
      <p>Batch command to submit jobs on the cluster: <br>
        <input type="text" name="cluster.batchCommand"><br>
      </p>
      <p>Job script header to define # of nodes, # of cores, RAM, walltime, etc.<br>
        <input type="text" name="cluster.jobHeader"><br></p>
      <p>Number of classifier threads (Integer value passed to ClassifierModule if it has a number of threads parameter.
        If undefined, script.numThreads will be used instead because classifiers often
        require more resources than other modules.):
        <input type="number" name="cluster.numClassifierThreads"><br></p>
      <p>Run java as script module: (If Y, each JavaModule will instantiate a clone of the
        application in direct mode on a job node via a single worker script to avoid overworking
        the head node where BioLockJ is deployed):<b>
        <input type="radio" name="cluster.runJavaAsScriptModule" value="Y"> Yes<br>
        <input type="radio" name="cluster.runJavaAsScriptModule" value="N"> No<br></p>
      <p>Validate cluster parameters: (If Y, validate
        cluster.jobHeader/cluster.classifierHeader "ppn:" or "procs:"
        value matches script.numThreads/cluster.numClassifierThreads)<br>
        <input type="radio" name="cluster.validateParams" value="Y"> Yes<br>
        <input type="radio" name="cluster.validateParams" value="N"> No<br></p>
      <p>Cluster modules to load before execution: <br>
        <input type="text" name="cluster.modules"><br></p>
    </form>
  </div>


  <div id="scriptTab" class="tabcontent">
    <form id="script">
      <p>Exit program if errors in script occur: (script.exitOnError)<br>
        <input type="radio" name="exitOnError" value="Y"> Yes<br>
        <input type="radio" name="exitOnError" value="N"> No (failures will be logged to failure directory) <br></p>
      <p>Number of sequence files to process per script:<br>
        <input type="number" name="batchSize"><br></p>
      <p>Command to grant script execute permissions:
        <input type="text" name="scriptPermissions" value="chmod 774"><br></p>
      <p>Number of threads:
        <input type="text" name="numThreads"><br></p>
        scriptTimeout

        scriptDefaultHeader
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="reportTab" class="tabcontent">
    <form id="report">
      <p>Include number of hits in output: <br>
        <input type="radio" name="numHits" value="Y"> Yes<br>
        <input type="radio" name="numHits" value="N"> No<br></p>
      <p>Add number of reads to output: <br>
        <input type="radio" name="numReads" value="Y"> Yes<br>
        <input type="radio" name="numReads" value="N"> No<br></p>
      <p>Metadata attributes for R script to add to statistical models:<br>
        <input type="text" name="attributes">
        <br></p>
      <p>Min threshold for OTU counts (<i>below which they will be ignored</i>):<br>
        <input type="text" name="minOtuCount"><br></p>
      <!--  <p>Delimiter to sepearte genus and species:<br>
        <input type="text" name="emptySpaceDelim"><br></p>-->
      <p>Taxonomic levels included in report:<br>
        <input type="checkbox" name="taxonomyLevels" value="domain">Domain<br>
        <input type="checkbox" name="taxonomyLevels" value="phylum">Phylum<br>
        <input type="checkbox" name="taxonomyLevels" value="class">Class<br>
        <input type="checkbox" name="taxonomyLevels" value="order">Order<br>
        <input type="checkbox" name="taxonomyLevels" value="family">Family<br>
        <input type="checkbox" name="taxonomyLevels" value="genus">Genus<br>
        <input type="checkbox" name="taxonomyLevels" value="species">Species<br>
      </p>
      <p>Report log base: <br>
        <input type="radio" name="logBase" value="10">10<br>
        <input type="radio" name="logBase" value="e">e<br></p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  min threshold  (total number) vs min count (per sample)

  <div id="mailTab" class="tabcontent">
    <form id="mail">
      <p>Encrypted E-mail password:<br>
        <i>(Encrypted password from email.from account.  If BioLockJ is passed a 2nd parameter
          (in addition to the config file), the 2nd parameter should be the clear-text password.
          The password will be encrypted and stored in the prop file for future use.)</i><br>
        <strong>WARNING: Base64 encryption is only a trivial roadblock for malicious users.
          This functionality is intended merely to keep clear-text passwords out of the
          configuration files and should only be used with a disposable email.from account.</strong>
        <input type="text" name="encryptedPassword"><br>
        <p>Notification emails sent from this account, provided email.encryptedPassword is valid:
          <input type="text" name="from"><br>
          <p>Comma-separated email recipients list:
            <input type="text" name="to"><br>
          </p>
          <p>Email SMTP Host:<br>
            <input type="text" name="smtp.host"><br>
          </p>
          p>Email SMTP port:<br>
          <input type="text" name="smtp.port"><br>
        </p>
        <p>Set the SMTP authorization property.<br>
          <input type="radio" name="smtp.auth" value="Y">Yes<br>
          <input type="radio" name="smtp.auth" value="N">No<br></p>
        <p>Set the SMTP start TLS property.<br>
          <input type="radio" name="smtp.starttls.enable" value="Y">Yes<br>
          <input type="radio" name="smtp.starttls.enable" value="N">No<br></p>
        <button onclick="saveTab(this.parentNode.id)">Save</button>
        <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="rTab" class="tabcontent">
    <form id="r">
      <p>If Y, will generate R Script log files:<br>
        <input type="radio" name="debug" value="Y">Yes<br>
        <input type="radio" name="debug" value="N">No <br>
      </p>
      <p>R script adds nominal statistical models for the listed metadata attributes. (Comma seperated list)<br>
        <input type="text" name="nominalData"><br>
      </p>
      <p>R script adds numeric statistical models for the listed metadata attributes. (Comma seperated list)<br>
        <input type="text" name="numericData"><br>
      </p>
      <p>Rare OTU threshold: <br>
        <i>If >1, R will filter OTUs below # provided.  If <1, R will treat # as percentage and ignore OTUs not found in that percentage of table rows in each taxa level.</i>
        <input type="number" name="rareOtuThreshold"><br>
      </p>
      <p><strong>Attributes Filter</strong><br>List of attributes for r to report on:
        <input type="text" name="filterAttributes"><br>
      </p>
      <p>List of logical operators to apply to attributes filter:
        <input type="text" name="filterOperators"><br>
      </p>
      <p>Filter values: <br><i><An orrdered list of values to compare with attributes filter)</i>
        <input type="text" name="filterValues"><br>
      </p>
      <p>Rows with NA values for attributes ignored in R script:
        <input type="text" name="filterNaAttributes">
      </p>
      <p>Number of breaks in P-value histograms output by R script:
        <input type="text" name="numHistogramBreaks">
      </p>
      <p>P-value adjustment scope: Used to set the p.adjust "n" parameter for how many simultaneous p-value calculations:<br>
        <input type="radio" name="adjPvalsScope" value="GLOBAL">GLOBAL<br>
        <input type="radio" name="adjPvalsScope" value="LOCAL">LOCAL<br>
        <input type="radio" name="adjPvalsScope" value="TAXA">TAXA<br>
        <input type="radio" name="adjPvalsScope" value="ATTRIBUTE">ATTRIBUTE<br>
      </p>
      <p>This is the base color used for labels & headings in the PDF report:<br>
        <input type="text" name="baseColor">
      </p>
      <p>This color is used to highlight significant attributes:<br>
        <input type="text" name="highlightColor">
      </p>
      <p>Sets the color of scatterplot points:<br>
        <input type="text" name="pointColor">
      </p>
      <p>Sets the p.adjust "method" parameter:<br><a href=https://stat.ethz.ch/R-manual/R-devel/library/stats/html/p.adjust.html>More info</a><br>
        <input type="radio" name="pAdjMethod" value="holm">Holm<br>
        <input type="radio" name="pAdjMethod" value="hochberg">Hochberg<br>
        <input type="radio" name="pAdjMethod" value="hommel">Hommel<br>
        <input type="radio" name="pAdjMethod" value="bonferroni">Honferroni<br>
        <input type="radio" name="pAdjMethod" value="BH">Benjamini & Hochberg (BH)<br>
        <input type="radio" name="pAdjMethod" value="BY">Benjamini & Yekutieli (BY)<br>
        <input type="radio" name="pAdjMethod" value="fdr">FDR<br>
        <input type="radio" name="pAdjMethod" value="none">none<br>
      </p>
      <p>Sets the format used in R sprintf() function:<br>
        <input type="text" name="pValFormat"><br>
      </p>
      <p>Set p-value cutoff used to assign label r.highlightColor:<br>
        <input type="number" name="pvalCutoff"><br>
      </p>
      <p>Set PDF report plot label maximum width:<br>
        <input type="number" name="plotWidth"><br>
      </p>
      <p>Set PDF report plot label position: (r.topRowAdj)<br>
        <input type="number" name="topRowAdj"><br>
      </p>
      <p>Set PDF report plot label position: (r.bottomRowAdj)<br>
        <input type="number" name="bottomRowAdj"><br>
      </p>
      <p>Set PDF report plot label position: (r.leftColAdj)<br>
        <input type="number" name="leftColAdj"><br>
      </p>
      <p>Set PDF report plot label position: (r.rightColAdj)<br>
        <input type="number" name="rightColAdj"><br>
      </p>
      <p>Set the size of histogram & boxplot headings in PDF reports<br>
        <input type="number" name="headingLabelSize"><br>
      </p>
      <p>Set the size of metadata attribute names in PDF reports<br>
        <input type="number" name="attributeLabelSize"><br>
      </p>
      <p># minutes before R Script times out:<br>
        <input type="number" name="timeout"><br>
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="exeTab" class="tabcontent">
    <p>If the program executable isn't already in your classpath ($PATH), add the executable path below:
    </p>
    <!--Make it so default commands are hints that will not be saved.-->
     <form id="exe">
      <p>Define executable awk command if default "awk" is not included in your $PATH:
        <input type="text" name="exe.awk">
      </p>
      <p>Define executable bowtie2 command if default "bowtie2" is not included in your $PATH:
        <input type="text" name="exe.bowtie2">
      </p>
      <p>Optional bowtie2 parameters used by SlimmClassifier:
        <input type="text" name="exe.bowtie2Params">
      </p>
      <p>Classifier executable command:
        <input type="text" name="exe.classifier">
      </p>
      <p>Optional classifier parameters,
        excluding parameters generated by BioLockJ (input files,
        output files, numThreads, etc.):
        <input type="text" name="exe.classifierParams">
      </p>
      <p>Define executable gzip command if default "gzip" is not included in your $PATH:
        <input type="text" name="exe.gzip">
      </p>
      <p>Define executable java command if default "java" is not included in your $PATH
        <input type="text" name="exe.java">
      </p>
      <p>Define executable pear command if default "pear" is not included in your $PATH
        <input type="text" name="exe.pear">
      </p>
      <p>Optional PEAR parameters:
        <input type="text" name="exe.pearParams">
      </p>
      <p>Define executable python command, if default "python" is not included in your $PATH:
        <input type="text" name="exe.python">
      </p>
      <p>Define executable Rscript command, if default "Rscript" is not included in your $PATH:
        <input type="text" name="exe.Rscript">
      </p>
      <p>Define executable samtools command, if default "samtools" is not included in your $PATH
        <input type="text" name="exe.samtools">
      </p>
      <p>Define executable vsearch command, if default "vsearch" is not included in your $PATH:
        <input type="text" name="exe.vsearch">
      </p>
      <p>Optional vsearch parameters:
        <input type="text" name="exe.vsearchParams">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
      <button type="button" class="nextTab">Save and move to next tab</button>
    </form>
  </div>

  <div id="rdpTab" class="tabcontent">
    <form id="rdp">
      <p>Required RDP minimum threshold score for valid OTUs:
        <input type="text" name="minThresholdScore">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>

  <div id="qiimeTab" class="tabcontent">
    <form id="qiime">
      <p>Alpha diverisity metrics: <br> Options listed online:
        <a href="http://scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html">
          scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html</a>.<br>
        <input type="text" name="alphaDiversityMetrics">
      </p>
      <p>Remove chimeras after open or de novo OTU picking:<br>
        <input type="radio" name="removeChimeras" value="Y">Yes<br>
        <input type="radio" name="removeChimeras" value="N">No <br>
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>

  <div id="krakenTab" class="tabcontent">
    <form id="kraken">
      <p>Path to Kraken database:
        <input type="text" name="kraken.db">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a id="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>

  <div id="slimmTab" class="tabcontent">
    <form id="slimm">
      <p>Path to SLIMM database:
        <input type="text" name="slimm.db">
      </p>
      <p>Path the bowtie2 reference genome index:
        <input type="text" name="refGenomeIndex">
      </p>
      <button onclick="saveTab(this.parentNode.id)">Save</button>
      <button class="createDownload">Create configuration file for download</button>
      <a class="downloadlink" class="hidden">Download configuration file to default directory</a>
    </form>
  </div>
</div>
<div class="banner">
  <!--Working on logo here-->
  <?xml version="1.0" standalone="No"?>
  <svg width="150" height="75" version="1.1" xmlns="http://www.w3.org/2000/svg">
    <!--rounded edges rect-->
    <rect x="60" y="20" rx="10" ry="10" width="30" height="30" stroke="black" fill="transparent" stroke-width="4"/>
    <!--<ellipse cx="75" cy="75" rx="20" ry="5" stroke="red" fill="transparent" stroke-width="5"/>-->
    <path d="M60,23 Q70,5  80,5 T90,23" fill="none" stroke="rgb(243,144,29)" stroke-width="2"/>
    <path d="M60,23 Q65,3  75,5 T90,23" fill="none" stroke="rgb(0,112,60)" stroke-width="2"/>
    <text x="69" y="43" dx="0" font-family="Times">J</text>
    <!--<path d="M70,30 L80,30  M75,30 L75,35 " fill="none" stroke="rgb(0,112,60)" stroke-width="2"/>-->
  </svg>
</div>
<script type="text/javascript" src="config_generator.js"></script>
</body>
</html>
